src/ServiceControl.Monitoring.Data/RawDataReporter.cs

```diff
@@ -67,17 +80,23 @@ public void Start()
                         await Task.Delay(singleSpinningTime).ConfigureAwait(false);
                     }
 
-                    await Consume().ConfigureAwait(false);
+                    if (consumers.Count >= MaxParallelConsumers)
+                    {
+                        var task = await Task.WhenAny(consumers).ConfigureAwait(false);
+                        consumers.Remove(task);
+                    }
+
+                    consumers.Add(Consume());
```

Comment:
I don't think this will work. AFAIK the RingBuffer is designed for single reader. With this approach there can be more than one reader reading from he buffer.

I would propose to keep a single reader but one that would read from the buffer until it gets empty (possibly resulting in more than one onChunk call with each chunk capped at maxFlushSize).

Reply:
@tmasternak I think it will ;-) RingBuffer is designed for a single reader. We call `Consume` that reads from one thread and add `sender(body)` to the task list. So consumption is done by a single thread, the sending part is asynchronous but it's done on the data copied  to `byte[]`.